src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagLinkedList.cs

```diff
@@ -149,6 +150,121 @@ public void AddFront(T value)
         public DiagEnumerator<T> GetEnumerator() => new DiagEnumerator<T>(_first);
         IEnumerator<T> IEnumerable<T>.GetEnumerator() => GetEnumerator();
         IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        private static void ActivityLinkToString(ref ActivityLink al, ref ValueStringBuilder vsb)
+        {
+            ActivityContext ac = al.Context;
+
+            vsb.Append("(");
+            vsb.Append(ac.TraceId.ToHexString());
+            vsb.Append(", ");
+            vsb.Append(ac.SpanId.ToHexString());
+            vsb.Append(", ");
+            vsb.Append(ac.TraceFlags.ToString());
+            vsb.Append(", ");
+            vsb.Append(ac.TraceState ?? "null");
+            vsb.Append(", ");
+            vsb.Append(ac.IsRemote ? "true" : "false");
+
+            if (al.Tags is not null)
+            {
+                vsb.Append(", [");
+                string sep = "";
+                foreach (KeyValuePair<string, object?> kvp in al.EnumerateTagObjects())
+                {
+                    vsb.Append(sep);
+                    vsb.Append(kvp.Key);
+                    vsb.Append(": ");
+                    vsb.Append(kvp.Value?.ToString() ?? "null");
+                    sep = ", ";
+                }
+
+                vsb.Append("]");
+            }
+            vsb.Append(")");
+        }
+
+        private static void ActivityEventToString(ref ActivityEvent ae, ref ValueStringBuilder vsb)
+        {
+            vsb.Append("(");
+            vsb.Append(ae.Name);
+            vsb.Append(", ");
+            vsb.Append(ae.Timestamp.ToString("o"));
+
+            if (ae.Tags is not null)
+            {
+                vsb.Append(", [");
+                string sep = "";
+                foreach (KeyValuePair<string, object?> kvp in ae.EnumerateTagObjects())
+                {
+                    vsb.Append(sep);
+                    vsb.Append(kvp.Key);
+                    vsb.Append(": ");
+                    vsb.Append(kvp.Value?.ToString() ?? "null");
+                    sep = ", ";
+                }
+
+                vsb.Append("]");
+            }
+            vsb.Append(")");
+        }
+
+        public override string ToString()
```

Comment:
@tarekgh - I take it the BCL is comfortable with changing the behavior of ToString() on a type that was previously released? I'm not pushing back but wasn't sure what the general stance was.

Reply:
This is an internal type. So, no public API change. There is a behavior change but this is what we are trying to fix here.

Reply:
I don't care if *Enumerate works. From my POV *Enumerate was just a means to an end - if we have another way of getting the result we like I'm happy. I suspect there are only a couple of consumers of this data in the world so the bar for clarify/consistency is pretty low :)